File: code/0876 Middle of the Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        A = head
        B = A
        while A.next and A.next.next:
            B = B.next
            A = A.next.next
        if A.next is None:
            return B
        else:
            return B.next

File: code/0876 Middle of the Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* middleNode(struct ListNode* head) {
    if(head->next==NULL){
        return head;        // We need to consider this case
    }
    struct ListNode *A=head,*B=head->next;
    while(B->next!=NULL){
        if(B->next->next==NULL){
            break;
        }
        // The main idea is that A moves one step while B moves two.
        A=A->next;
        B=B->next->next;
    }
    return A->next;
}

